design-documents/checkout/alternative-checkout-flow.md

```diff
@@ -0,0 +1,63 @@
+# Alternative checkout flow
+
+In scope of work on GraphQL and storefront APIs we have an opportunity to improve design and features of storefront checkout.
+
+The purpose of this document is to discuss possible alternatives to current Magento checkout flow. 
+
+The cart is just a container for the items the user wants to purchase. In the proposed flow the cart is created as soon as user adds product to cart. The data from cart entity should be enough to render minicart and cart pages. Taxes and other adjustments are not calculated at these steps.
+
+The quote, on the other hand, contains a full break down of all adjustments calculated. It provides the user with the total he has to pay for the items in the cart.
+
+It is possible to split cart items into separate quotes. This can be done based on shipping addresses or shipping sources.
+
+Each quote is used to create a separate order. Multiple payment methods (with independent billing address each) can be selected during each order creation.
+
+Cart will depend on Catalog. Quote will depend on Cart and must not depend on Catalog. Order will depend on Quote and must not depend on Cart or Catalog.
+
+  1. When Quote is created?
+     * For physical products on Review & Payments step
+     * For virtual products billing address has to be entered first
+  1. Cart properties:
+     * Line Items:
+       * SKU
+       * Selected options (custom/configurable)
+       * Quantity
+       * Regular price
+       * Price
+  4. Applicable cart rules calculator arguments
+     * Cart
+     * Addresses
+     * Date & Time (admin preview)
+  2. Quote factory arguments: Quote
+     * !Line Items
+     * Dimensions (weight & size) (based on LineItems)
+     * Shipping (requited for physical products)
+       * Address
+       * Selected Shipping Method
+     * Billing address (required for virtual products). Need use cases
+     * Coupons
+     * Gift cards
+     * Store credit
+     * Cart rules (calculated by Applicable catalog rules calculator)
+     * Customer
+  2. Totals calculator: Totals
```

Comment:
in the video of the architecture meeting where this proposal was discussed it was mentioned that totals are entities. Was that meant literally?
From my point of view the totals are a function of the quote.
The results of the calculation may be stored for performance reasons, but a total does not have any identity over time.
Also, I didn't understand the point about how this new proposal allows for sorting of total calculation.
It sounded as if that is not possible in the current implementation. However, it certainly is possible.

The downside of the current implementation is that each total aggregates it's results on the quote items and the quote. Shared mutable state is always tricky. However, it does allow for many customizations where one total calculation can change the value of a previous one. After listening to the discussion, I'm not sure if the total results where supposed to be immutable or not. If they are, how would such customizations be possible?

Reply:
> in the video of the architecture meeting where this proposal was discussed it was mentioned that totals are entities. Was that meant literally?

The idea is to have a separate immutable extensible entity for totals which can be used in other entities like order, invoice, credit memo because there are issues with the calculation of their total.

> I didn't understand the point about how this new proposal allows for sorting of total calculation.

There will be some kind of class which will check the configuration and create to a composite of totals in the order according to configuration or applied rules (like cart rules).

> It sounded as if that is not possible in the current implementation.

Actually no. The point was that the current mechanism is hard to understand because you can see the only configuration in Admin Panel and logic is smashed across the code. The main idea is making it more visible (even provide some kind of visualization in Admin Panel) and have only one place in the code.

> After listening to the discussion, I'm not sure if the total results where supposed to be immutable or not. If they are, how would such customizations be possible?

Totals should be immutable. Each totals calculator uses the result of previous totals calculator, any extension/customization can add own calculator to the list.